EIPS/pairings.md

```diff
@@ -0,0 +1,131 @@
+
+Note that fixing these parameters will in no way limit the use-cases for zkSNARKs, it will even allow for incorporating some advances in zkSNARK research without the need for a further hard fork.
+
+Pairing functions can be used to perform a limited form of multiplicatively homomorphic operations, which are necessary for current zkSNARKs. This precompile can be used to run such computations within the block gas limit. This precompiled contract only specifies a certain check, and not an evaluation of a pairing function. The reason is that the codomain of a pairing function is a rather complex field which could provide encoding problems and all known uses of pairing function in zkSNARKs only require the specified check.
+
+## Specification
+
+Add a precompiled contracts for a bilinear function on groups on the elliptic curve "alt_bn128". We will define the precompiled contract in terms of a discrete logarithm. The discrete logarithm is of course assumed to be hard to compute, but we will give an equivalent specification that makes use of elliptic curve pairing functions which can be efficiently computed below.
+
+Address: 0x8
+
+For a cyclic group `G` (written additively) of prime order q let `log_P: G -> F_q` be the discrete logarithm on this group with respect to a generator `P`, i.e. `log_P(x)` is the integer `n` such that `n * P = x`.
+
+The precompiled contract is defined as follows, where the two groups `G_1` and `G_2` and their generators `P_1` and `P_2` are defined below (they have the same order `q`):
+
+```
+Input: (a1, b1, a2, b2, ..., ak, bk) from (G_1 x G_2)^k
+Output: If the length of the input is incorrect or any of the inputs are not elements of
+        the respective group or are not encoded correctly, the call fails.
+        Otherwise, return one if
+        log_P1(a1) * log_P2(b1) + ... + log_P1(ak) * log_P2(bk) = 0
+        (in F_q) and zero else.
+```
+
+Note that k is determined from the length of the input. k == 0 is valid and results in returning one.
+
+### Definition of the groups
+
+The groups `G_1` and `G_2` are cyclic groups of prime order `q` on the elliptic curve `alt_bn128` defined by the curve equation
+`Y^2 = X^3 + 3`.
+
+The group `G_1` is a cyclic group on the above curve over the field `F_p` with `p = 21888242871839275222246405745257275088696311157297823662689037894645226208583` with generator `P1 = (1, 2)`.
+
+The group `G_2` is a cyclic group on the same elliptic curve over a different field `F_p^2 = F_p[i] / (i^2 + 1)` (p is the same as above) with generator
+```
+P2 = (
+  11559732032986387107991004021392285783925812861821192530917403151452391805634 * i +
+  10857046999023057135944570762232829481370756359578518086990519993285655852781,
+  4082367875863433681332203403145435568316851327593401208105741076214120093531 * i +
+  8495653923123431417604973247489272438418190587263600148770280649306958101930
+)
+```
```

Comment:
Let `G_3` be the cyclic subgroup generated by `P3`, where `P3` is very similar to `P2` but uses `-i` instead of `i`.  Is it the case that `G_3` has the same order as `G_2`?  If that's the case, the current input checking method checks `G_2 ∪ G_3 ` I guess.